crates/ruff/src/rules/flake8_pyi/rules/duplicate_union_member.rs

```diff
@@ -0,0 +1,79 @@
+use rustc_hash::FxHashSet;
+use rustpython_parser::ast::{Expr, ExprKind, Operator};
+
+use ruff_diagnostics::{AlwaysAutofixableViolation, Diagnostic, Edit};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::comparable::ComparableExpr;
+use ruff_python_ast::helpers::unparse_expr;
+use ruff_python_ast::types::Range;
+
+use crate::checkers::ast::Checker;
+use crate::registry::AsRule;
+
+#[violation]
+pub struct DuplicateUnionMember {
+    pub duplicate_name: String,
+}
+
+impl AlwaysAutofixableViolation for DuplicateUnionMember {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Duplicate union member `{}`", self.duplicate_name)
+    }
+
+    fn autofix_title(&self) -> String {
+        format!("Remove duplicate union member `{}`", self.duplicate_name)
+    }
+}
+
+/// PYI016
+pub fn duplicate_union_member(checker: &mut Checker, expr: &Expr) {
```

Comment:
I tweaked this to use an inner helper, so that the caller doesn't have to worry or know about (e.g.) creating a hash set for internal use.